File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {

        if(root == null){

            return;

        }

        

        TreeNode left = root.left;

         TreeNode right = root.right;

        

        flatten(left);

        if(left != null){

            root.right = left;

            root.left = null;

            

            

            TreeNode curr = root;

            while(curr.right != null){

                curr = curr.right;

            }

            curr.right = right;

            

        }

        

        flatten(root.right);

        

    }
}

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.cpp
class Pair{
    
    public:
      TreeNode *first,*last;
    
};

class Solution {
    
    Pair helper(TreeNode *root)
    {
        if(root==NULL)
        {
            Pair ans;
            ans.first=NULL;
            ans.last=NULL;
            return ans;
        }
        Pair Leftans=helper(root->left);
        Pair Rightans=helper(root->right);
        
        Pair ans;
        if(Leftans.first==NULL && Rightans.first==NULL)
        {
            ans.first=root;
            ans.last=root;
        }
        else if(Leftans.first==NULL)
        {
            root->right=Rightans.first;
            ans.first=root;
            ans.last=Rightans.last;
        }
        else if(Rightans.first==NULL)
        {
            root->right=Leftans.first;
            root->left=NULL;
            ans.first=root;
            ans.last=Leftans.last;
        }
        else
        {
            root->right=Leftans.first;
            root->left=NULL;
            Leftans.last->right=Rightans.first;
            ans.first=root;
            ans.last=Rightans.last;
        }
        
        return ans;
    }
    
public:
    void flatten(TreeNode* root) {
        
        Pair ans=helper(root);
        // return ans.first;
        
    }
};